Deprecate onReorder callback

navaronbracke opened the pull request:
This pull request deprecates the old `onReorder` callback of `ReorderableList` / `SliverReorderableList` / `ReorderableListView`, in favor of a new callback, `onReorderItem`, that does the index correction as mentioned in the docs at https://api.flutter.dev/flutter/widgets/ReorderCallback.html

I added tests to ensure that the old callback is preferred when both are provided and tests that assert that the new callback does the index correction correctly. I also noticed that some tests used empty callbacks for `onReorder`, those have been updated to use wildcards in a consistent manner.

I did not opt into making a design doc, since the change itself is very small, that will also be apparent in the migration guide that I am going to write.

Fixes https://github.com/flutter/flutter/issues/127901
Fixes https://github.com/flutter/flutter/issues/169878

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

Migration guide: https://github.com/flutter/website/pull/13012

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

navaronbracke committed:
deprecate onReorder in Material ReorderableList

navaronbracke committed:
deprecate onReorder in widgets ReorderableList/SliverReorderableList

navaronbracke committed:
update examples to use the new callback

navaronbracke committed:
update SliverReorderableList to use the new callback

navaronbracke committed:
update Material ReorderableList test

navaronbracke committed:
update widgets SliverReorderableList / ReorderableList test

navaronbracke committed:
add test for preference of onReorder if provided in SliverReorderableList / ReorderableList

navaronbracke committed:
add test for preference of onReorder if provided in ReorderableListView / ReorderableListView.builder

navaronbracke committed:
add test for onReorderItem in SliverReorderableList / ReorderableList

navaronbracke committed:
add test for onReorderItem in ReorderableListView

navaronbracke committed:
fix docs in ReorderCallback

navaronbracke committed:
fix typo in deprecation message

dkwingsmt reviewed (commented):
Generally looking good. Thank you!

navaronbracke commented on packages/flutter/lib/src/widgets/reorderable_list.dart:
```diff
@@ -151,7 +143,13 @@ class ReorderableList extends StatefulWidget {
+      'This feature was deprecated after v3.38.0-0.2.pre.',
+    )
+    this.onReorder,
+    required this.onReorderItem,
```
Yes!  I forgot to rectify this. I'll update the behavior once I resynced with upstream (this PR was from before the flutter_lints merge)

navaronbracke commented on packages/flutter/test/material/reorderable_list_test.dart:
```diff
@@ -2032,6 +2023,274 @@ void main() {
   });
 
+  testWidgets(
+    'ReorderableListView prefers old onReorder callback if provided along with onReorderItem callback',
```
I will add a different test, that verifies the assert instead.

navaronbracke committed:
sync analysis_options with upstream flutter/flutter

navaronbracke committed:
fix missing types

navaronbracke committed:
fix type annotation

navaronbracke committed:
fix doc macro

navaronbracke committed:
update deprecation version

navaronbracke committed:
make onReorderItem nullable; add assert

navaronbracke committed:
repurpose old test to test old behavior

navaronbracke committed:
fix argument name

navaronbracke committed:
add assert tests

navaronbracke committed:
reuse reorder method for semantics action

navaronbracke commented on packages/flutter/lib/src/widgets/reorderable_list.dart:
```diff
@@ -1050,9 +1090,18 @@ class SliverReorderableListState extends State<SliverReorderableList>
+        }
+
+        widget.onReorderItem.call(oldIndex, newIndex);
       }
```
Fixed.

navaronbracke commented:
I updated the PR, so that the new parameter is nullable. I also added new tests for the assert and repurposed the old tests that I wrote, for testing if the old callback still works (if a user were to provide it).

Also cleaned up one of Gemini's suggestions, because some code was duplicated.

dkwingsmt commented on packages/flutter/lib/src/widgets/reorderable_list.dart:
```diff
@@ -207,8 +210,24 @@ class ReorderableList extends StatefulWidget {
+  ///
+  /// If `oldIndex` is before `newIndex`, removing the item at `oldIndex` from the
+  /// list will reduce the list's length by one. Implementations will need to
+  /// account for this when inserting before `newIndex`.
```
I find that it's not immediately obvious how existing code should be migrated, which is especially inappropriate for an essential method like this. I propose the following addition. Feel free to change as you see fit.
```suggestion
  /// account for this when inserting before `newIndex`.
  ///
  /// This callback has been deprecated in favor of [onReorderItem], which
  /// simplifies the reordering logic by automatically handling index
  /// adjustments. To migrate, remove the manual adjustment of `newIndex` when
  /// items are moved downward in the list. For example, replace:
  /// ```dart
  /// onReorder: (oldIndex, newIndex) {
  ///   if (newIndex > oldIndex) {
  ///     newIndex -= 1;
  ///   }
  ///   // Handle reordering...
  /// }
  /// ```
  ///
  /// With:
  /// ```dart
  /// onReorderItem: (oldIndex, newIndex) {
  ///   // Handle reordering...
  /// }
  /// ```
```

dkwingsmt reviewed (approved):
LGTM with one doc suggestion.

Sorry for the late review!

dkwingsmt committed:
Merge branch 'master' into deprecate_onreorder_callback

victorsanni commented on packages/flutter/lib/src/material/reorderable_list.dart:
```diff
@@ -114,6 +120,11 @@ class ReorderableListView extends StatefulWidget {
          children.every((Widget w) => w.key != null),
          'All children of this widget must have a key.',
        ),
+       assert(
```
This assert basically requires `onReorderItem` for new instances of this class, but that's not reflected in the API contract. Do we need to steer developers to always provide `onReorderItem` through the class API docs (or assert message)? I know requiring `onReorderItem` will be a breaking change, but I'm thinking of a future where the deprecated property gets removed and a developer finds out the hard way that a property which isn't marked as required actually is.

Maybe there is prior art for this?

navaronbracke replied:
Yes, this is what https://github.com/flutter/flutter/pull/178242#discussion_r2692534620 was also about.
I'm going to add a migration guide on the website (still TBD) and add a doc comment about it.

I can update the assert message to nudge towards the replacement.
The idea is that when the deprecated property is removed, the new callback becomes nonnull and required. (in the same change that removes the old callback)

navaronbracke committed:
update deprecation version; update assert message; add migration sample in docs

dkwingsmt committed:
Merge branch 'master' into deprecate_onreorder_callback

dkwingsmt committed:
Merge branch 'master' into deprecate_onreorder_callback

dkwingsmt reviewed (approved):
Thank you!

victorsanni committed:
Merge branch 'master' into deprecate_onreorder_callback
